vector/test.c

```diff
@@ -0,0 +1,94 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include "array.h"
+#include "test.h"
+
+void test_vector_create() {
+    array *arr = vector_create();
+    ASSERT(arr->capacity == init_capacity);
+    ASSERT(arr->size == 0);
+
+    vector_free(arr);
+}
+
+void test_vector_append() {
+    array *arr = vector_create();
+    vector_append(arr, 5);
+    ASSERT(arr->array[0] == 5);
+    ASSERT(arr->size == 1);
+
+    vector_free(arr);
+}
+
+void test_vector_pop() {
+    array *arr = vector_create();
+    vector_append(arr, 5);
+    vector_append(arr, 4);
+    vector_append(arr, 3);
+    data_t popped_el = 0;
+    vector_pop(arr, &popped_el);
+    ASSERT(popped_el == 3);
+    ASSERT(arr->size == 2);
+
+    vector_free(arr);
+}
+
+void test_vector_create_size() {
+    array *arr = vector_create_size(36);
+    ASSERT(arr->capacity == 36);
+    ASSERT(arr->size == 0);
+    free(arr);
+}
+
+void test_vector_insert() {
+    array *arr = vector_create();
+    vector_append(arr, 5);
+    vector_append(arr, 4);
+    vector_append(arr, 3);
+    vector_insert(arr, 100, 2);
+    ASSERT(arr->array[2] == 100);
+
+    vector_free(arr);
+}
+
+void test_vector_remove() {
+    array *arr = vector_create();
+    vector_append(arr, 5);
+    vector_append(arr, 4);
+    vector_append(arr, 3);
+    vector_insert(arr, 100, 2);
+    vector_remove(arr, 2);
+    ASSERT(arr->array[2] == 3);
+
+    vector_free(arr);
+}
+
+void test_vector_copy() {
+    array *arr = vector_create();
+    vector_append(arr, 0);
+    vector_append(arr, 3);
+    vector_append(arr, 5);
+    array *copy_arr = vector_copy(arr);
+    ASSERT(copy_arr->array[0] == 0);
+    ASSERT(copy_arr->array[1] == 3);
+    ASSERT(copy_arr->array[2] == 5);
+    ASSERT(copy_arr->size == 3);
+    ASSERT(copy_arr->capacity == 4);
+
+    vector_free(arr);
+    vector_free(copy_arr);
+}
+
+void test_vector_create_size_NULL_return() {
+    size_t max_size = (size_t) -1   ;
```

Comment:
Hmm... I was wrong. This can't be `SIZE_MAX` here. It must be `SIZE_MAX` divided by the size of data.

Because otherwise, valgrind says this:
```
==4874== Argument 'size' of function malloc has a fishy (possibly negative) value: -4
==4874==    at 0x4C2BBCF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==4874==    by 0x400A5F: vector_create_size (array.c:75)
==4874==    by 0x401340: test_vector_create_size_NULL_return (test.c:85)
==4874==    by 0x4007C9: main (test_main.c:11)
==4874== 
```